index.js

```
@@ -1,77 +1,184 @@
+    this.unwatch = watch(this.filename, this.onchange)
+
+    this.opened = true
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    if (!this.opened) await this._opening.catch(safetyCatch)
+
+    if (this.unwatch) {
+      await this.unwatch()
+      this.unwatch = null
     }
-    clear = st.isDirectory() ? watchDirectory(name, onchange) : watchFile(name, onchange)
-  })
+  }
+}
+
+class Watch {
+  constructor (filename, opts, onchange) {
+    this.filename = filename
+    this.opts = opts || {}
+    this.onchange = onchange
+
+    this.opened = true
+    this.closed = false
 
-  return function () {
-    if (stopped) return
-    stopped = true
-    if (clear) {
-      clear()
+    this._closing = null
+
+    this.watcher = fs.watch(this.filename, this.opts, this.onchange)
+    this.watcherClosed = false
+
+    this.watcher.on('error', noop)
+    this.watcher.on('close', () => {
+      this.watcherClosed = true
+    })
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    let onclose = null
+
+    if (!this.watcherClosed) {
+      onclose = new Promise(resolve => this.watcher.on('close', resolve))
     }
+
+    this.watcher.close()
+    await onclose
+
+    this.watcher = null
   }
 }
 
 function watchFile (filename, onchange) {
-  var prev = null
-  var prevTime = 0
+  let prev = null
+  let prevTime = 0
+  let actives = 0
+
+  let closed = false
+  let close = null
+  const closing = new Promise(resolve => { close = resolve })
+
+  const watcher = new Watch(filename, null, function () {
+    actives++
 
-  var w = fs.watch(filename, function () {
     fs.lstat(filename, function (_, st) {
```

Not related to this PR, but curious why we ignore the err cb here

---

I had the same doubt, let's say it's watching a file then it gets deleted, so this callback gets executed once and `st` will be undefined triggering a last change due the `same` helper as well

---

So ignoring the error here doesn't seem that bad at the end